modules/core/src/main/java/org/apache/ignite/internal/managers/encryption/CacheGroupReencryption.java

```diff
@@ -0,0 +1,583 @@
+
+            if (fut == null)
+                return false;
+
+            return fut.cancel();
+        }
+
+        /**
+         * @param t Throwable.
+         */
+        public void fail(Throwable t) {
+            cpFut.onDone(t);
+        }
+    }
+
+    /**
+     * Page store scanning task.
+     */
+    private class PageStoreScanTask extends GridFutureAdapter<Void> implements Runnable {
+        /** Partiion ID. */
+        private final int partId;
+
+        /** Cache group reencryption context. */
+        private final GroupReencryptionContext encrCtx;
+
+        /**
+         * @param encrCtx Cache group reencryption context.
+         * @param partId Partition ID.
+         */
+        public PageStoreScanTask(GroupReencryptionContext encrCtx, int partId) {
+            this.encrCtx = encrCtx;
+            this.partId = partId;
+        }
+
+        /** {@inheritDoc} */
+        @Override public synchronized boolean cancel() throws IgniteCheckedException {
+            return onDone();
+        }
+
+        /** {@inheritDoc} */
+        @Override public void run() {
+            try {
+                int grpId = encrCtx.groupId();
+
+                CacheGroupContext grp = ctx.cache().cacheGroup(grpId);
+
+                if (grp == null) {
+                    onDone();
+
+                    return;
+                }
+
+                PageMemoryEx pageMem = (PageMemoryEx)grp.dataRegion().pageMemory();
+                long metaPageId = pageMem.partitionMetaPageId(grpId, partId);
+                long state = ctx.encryption().getEncryptionState(grpId, partId);
+
+                int pageIdx = (int)(state >> 32);
+                int cnt = (int)state;
+
+                if (log.isDebugEnabled()) {
+                    log.debug("Partition reencryption is started [grpId=" + grpId +
+                        ", p=" + partId + ", remain=" + (cnt - pageIdx) + ", total=" + cnt + "]");
+                }
+
+                while (pageIdx < cnt) {
+                    synchronized (this) {
+                        ctx.cache().context().database().checkpointReadLock();
+
+                        try {
+                            if (isDone() || ctx.encryption().getEncryptionState(grpId, partId) == 0)
+                                break;
+
+                            int end = Math.min(pageIdx + batchSize, cnt);
+
+                            do {
+                                long pageId = metaPageId + pageIdx;
+
+                                pageIdx += 1;
+
+                                long page = pageMem.acquirePage(grpId, pageId);
+
+                                try {
+                                    // Can skip rewriting a dirty page if the checkpoint has been completed.
+                                    if (encrCtx.skipDirty() && pageMem.isDirty(grpId, pageId, page))
+                                        continue;
+
+                                    pageMem.writeLock(grpId, pageId, page, true);
```

Comment:
why it here?

Reply:
Could you describe the question in more detail?